src/libopensc/ctx.c

```diff
@@ -831,25 +831,50 @@ int sc_context_repair(sc_context_t **ctx_out)
 #ifdef USE_OPENSSL3_LIBCTX
 static int sc_openssl3_init(sc_context_t *ctx)
 {
+	const char *openssl_cnf_path = NULL;
+
 	ctx->ossl3ctx = calloc(1, sizeof(ossl3ctx_t));
-	if (ctx->ossl3ctx == NULL)
+	if (ctx->ossl3ctx == NULL) {
 		return SC_ERROR_OUT_OF_MEMORY;
+	}
+
 	ctx->ossl3ctx->libctx = OSSL_LIB_CTX_new();
 	if (ctx->ossl3ctx->libctx == NULL) {
 		return SC_ERROR_INTERNAL;
 	}
-	ctx->ossl3ctx->defprov = OSSL_PROVIDER_load(ctx->ossl3ctx->libctx,
-						    "default");
-	if (ctx->ossl3ctx->defprov == NULL) {
-		OSSL_LIB_CTX_free(ctx->ossl3ctx->libctx);
-		free(ctx->ossl3ctx);
-		ctx->ossl3ctx = NULL;
-		return SC_ERROR_INTERNAL;
+
+#ifndef _WIN32
+	openssl_cnf_path = getenv("OPENSSL_CONF");
+	if (!openssl_cnf_path) {
+		openssl_cnf_path = OPENSC_OPENSSL_CONF_PATH;
 	}
-	ctx->ossl3ctx->legacyprov = OSSL_PROVIDER_load(ctx->ossl3ctx->libctx,
-						       "legacy");
-	if (ctx->ossl3ctx->legacyprov == NULL) {
-		sc_log(ctx, "Failed to load OpenSSL Legacy provider");
+#endif
+
+	if (openssl_cnf_path != NULL && access(openssl_cnf_path, R_OK) == 0) {
```

Comment:
I think if the user explicitly specifies some configuration file, but we don't have proper access rights, then we want to error instead of continuing with the default provider.

Reply:
Good point. Added a warning when the file is configured, but not readable.